### PR TITLE
Added platform/toolchain.h toolchain abstraction layer via macros.

### PR DESCRIPTION
**Pull request by turon**
Added toolchain abstraction file so __otPACKED, __otWEAK, __otALIGNED, and similar constructs can work across disparate toolchains other than GCC, such as Keil, IAR, and SDCC.  Also created a unit test to enforce the assumptions of the more common set of the macros.

### CODE REVIEW

**Comment by jwhui** [comment]
I agree that OpenThread should be compatible with as many toolchains as possible.  Abstracting out the `packed` attribute is certainly useful.  However, `toolchain.h` goes far beyond what is needed for interfacing with the OpenThread library - the only attribute OpenThread relies on is `packed`.  Projects including OpenThread should not be bothered with worrying about all other macros that have no dependency.  If we only add the `packed` macros, then I think it is sufficient to add to `openthread-types.h`.

Specifically on `packed`, what about defining a single macro that supports both pre-pended, post-pended, and even inserted decorations?

```c++
#define PACK(decl) decl __attribute__((packed))

PACK(struct foo
{
    ...
});
```

**Comment by darconeous** [comment]
Also, double underscore prefix is the implementor reserved namespace. Since OpenThread is a library, we shouldn't be using it under any circumstances.

Identifiers like "OPENTHREAD_PACKED` or `OT_COMPFLAG_PACKED` would be more appropriate.

**Comment by darconeous** [comment]
I'm not a fan of the function-style macro that @jwhui suggested (Also, I'd prefer all stand-alone macros that could never possibly be implemented as anything else be in all caps), but I admit the only immediate fault I find with the approach is aesthetics.

And it confuses my own visual parsing.

**Comment by darconeous** [comment]
I think it might be useful to have other kinds of macros for helping the compiler check syntax, etc. (Can provide examples tomorrow morning)

**Comment by jwhui** [comment]
@darconeous , I'm all for enabling more compiler checks whenever we can.

For the packed macro, I am fine with not using a function-style macro - that would result in having something like `OT_COMPFLAG_PACKED_BEFORE` and `OT_COMPFLAG_PACKED_AFTER`.

**Comment by darconeous** [comment]
(Using @turon's `OT_TOOL_*` prefix in this post after an in-person discussion on the matter)

Given that the IAR seems to be the only compiler that enforces the use of the packed flag as a prefix (meaning that it's use as a prefix is more bug than feature), I am very sympathetic to @turon's asymmetrical "`OT_TOOL_PACKED_BEFORE`/`OT_TOOL_PACKED`" notation, since it allows for `OT_TOOL_PACKED_BEFORE` to possibly be retired, leaving `OT_TOOL_PACKED` to remain unchanged.

However, I think there is a credible argument that adding `OT_TOOL_PACKED_AFTER` encourages the programmer to realize that there is likely something that needs to come `OT_TOOL_PACKED_BEFORE`. It would be very easy for a programmer to simply ignore the `OT_TOOL_PACKED_BEFORE` macro and think that they are doing just fine with `OT_TOOL_PACKED`.

So after thinking about this more, I think having a `OT_TOOL_PACKED_BEFORE` and a `OT_TOOL_PACKED_AFTER` (and no `OT_TOOL_PACKED`) makes sense from the perspective of avoiding potential (and very hard to diagnose) bugs.

**Comment by jwhui** [comment]
There are a handful of compilers that place packed decorations before the declaration, including [KEIL](http://www.keil.com/support/man/docs/armcc/armcc_chr1359124968737.htm) and [Visual Studio](https://msdn.microsoft.com/en-us/library/2e70t5y1.aspx).

There are even compilers that place packed decorations before *and* after the declaration, such as [CodeWarrior](http://www.nxp.com/files/soft_dev_tools/doc/white_paper/CWPORTINGWP.pdf).  I noted that the existing `toolchain.h` file in this PR seems to include CodeWarrior support, but the packed defines are empty.  Have we tested all of the included compilers?  We should only include compiler support for those we have tested against.

**Comment by darconeous** [comment]
> We should only include compiler support for those we have tested against.

Agreed in general, but I don't think CodeWarrior actually supports the same "packed" concept that the others do: <https://gbatemp.net/threads/how-to-use-pack-codewarrior.133478/>

**Comment by turon** [comment]
Yes, Keil historically has looked similar to IAR.  Luckily however, they
added support for GCC __attribute__ in the past few years in addition to
the legacy method, so that is what toolchain.h is currently leveraging.  If
someone is using a much older version, we can add detection for that and
handle accordingly (based on demand).

CodeWarrior unfortunately uses #pragma, which can't be triggered with
#define generally.  I'll move CodeWarrior to "unsupported" for now.

Visual Studio also uses #pragma, but luckily has __pragma equivalent which
should be usable from a #define.   I'll look into adding Visual Studio
support and testing it.

So the plan is to resolve the comments, hone the macros to the supported
subset, and update the pull request.

I'm thinking the following naming:

OT_TOOL_PACKED_BEGIN        ///< after typedef, before struct
OT_TOOL_PACKED_END          ///< at the end of struct


Martin

_____________________________
Martin Turon  |  Nest Labs

**Comment by jwhui** [comment]
@turon, I see your latest update still includes a bunch of macros in addition to `OT_TOOL_PACKED_BEGIN` and `OT_TOOL_PACKED_END`.  I still don't see why we need to expose these macros in an installed, public header file.  OpenThread should not be concerned with project-specific compiler attributes that it does not use.

File: include/platform/toolchain.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+#elif defined(__SDCC)
+
+#define OT_TOOL_WEAK
+#define OT_TOOL_PACKED_BEGIN
+#define OT_TOOL_PACKED_END
+#define OT_TOOL_ALIGNED
+#define OT_TOOL_DEPRECATED
+
+#elif defined(__CODEWARRIOR)
```
Comment (jwhui) [inline include/platform/toolchain.h]: Just remove this if statement if we are not supporting it.

File: include/platform/toolchain.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+#endif
+
+// =========== TOOLCHAIN SELECTION : START ===========
+
+#if defined(__GNUC__) || defined(__CC_ARM)
+
+// https://gcc.gnu.org/onlinedocs/gcc/Common-Variable-Attributes.html
+// http://www.keil.com/support/man/docs/armcc/armcc_chr1359124973480.htm
+
+#define OT_TOOL_WEAK                     __attribute__((weak))
```
Comment (jwhui) [inline include/platform/toolchain.h]: Not used by OpenThread.

File: include/platform/toolchain.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+#define OT_TOOL_USED                     __attribute__((used))
```
Comment (jwhui) [inline include/platform/toolchain.h]: Not used by OpenThread.

File: include/platform/toolchain.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+#define OT_TOOL_UNUSED                   __attribute__((unused))
```
Comment (jwhui) [inline include/platform/toolchain.h]: Not used by OpenThread.

File: include/platform/toolchain.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+#define OT_TOOL_PACKED_BEGIN
+#define OT_TOOL_PACKED_END               __attribute__((packed))
+#define OT_TOOL_ALIGNED                  __attribute__((aligned))
```
Comment (jwhui) [inline include/platform/toolchain.h]: Not used by OpenThread.

File: include/platform/toolchain.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+#define OT_TOOL_DEPRECATED               __attribute__((deprecated))
```
Comment (jwhui) [inline include/platform/toolchain.h]: Not used by OpenThread.

**Comment by turon** [comment]
@jwhui  Yes, sorry, was in process of stripping out all but OT_TOOL_PACK_START / END.  
@darconeous suggested leaving OT_TOOL_DEPRECATED.

**Comment by jwhui** [comment]
@turon, thanks for the update.  I can see `OT_TOOL_DEPRECATED` being used by OpenThread in the future.

**Comment by turon** [comment]
I think it's ready.  The CI is blocked for some reason.  Anyone know how to kick it off again?

File: include/platform/toolchain.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file defines a toolchain abstraction layer through macros.
+ *
+ */
+
+#ifndef _TOOLCHAIN_H_
```
Comment (jwhui) [inline include/platform/toolchain.h]: Can we make the guard consistent with the other header files?

File: include/platform/toolchain.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+#endif
+
+// =========== TOOLCHAIN SELECTION : START ===========
+
+#if defined(__GNUC__) || defined(__CC_ARM)
+
+// https://gcc.gnu.org/onlinedocs/gcc/Common-Variable-Attributes.html
+// http://www.keil.com/support/man/docs/armcc/armcc_chr1359124973480.htm
+
+#define OT_TOOL_PACKED_BEGIN
```
Comment (jwhui) [inline include/platform/toolchain.h]: Missing doxygen.

File: include/platform/toolchain.h [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+#define OT_TOOL_PACKED_END                  __attribute__((packed))
```
Comment (jwhui) [inline include/platform/toolchain.h]: Missing doxygen.

File: include/platform/toolchain.h [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+#define OT_TOOL_DEPRECATED(symbol)          __attribute__((deprecated))
```
Comment (jwhui) [inline include/platform/toolchain.h]: Missing doxygen.

File: include/platform/toolchain.h [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+#else
+
+#error "Error: No valid Toolchain specified"
+
+#endif
+
+// =========== TOOLCHAIN SELECTION : END ===========
+
+#ifdef __cplusplus
+}
```
Comment (jwhui) [inline include/platform/toolchain.h]: Add `// extern "C"`

File: src/core/thread/lowpan.hpp [hunk 12]
```diff
@@ -421,10 +423,10 @@ class FragmentHeader
     {
         uint8_t mDispatchOffsetSize;
         uint16_t mSize;
-    } __attribute__((packed));
+    } OT_TOOL_PACKED_END;
```
Comment (jwhui) [inline src/core/thread/lowpan.hpp]: Missing a matching `OT_TOOL_PACKED_BEGIN`?

File: src/core/thread/mle_tlvs.hpp [hunk 13]
```diff
@@ -1313,13 +1332,14 @@ class AddressRegistrationEntry
     {
         uint8_t mIid[Ip6::Address::kInterfaceIdentifierSize];
         Ip6::Address mIp6Address;
-    } __attribute__((packed));
-} __attribute__((packed));
+    } OT_TOOL_PACKED_END;
```
Comment (jwhui) [inline src/core/thread/mle_tlvs.hpp]: Missing a matching `OT_TOOL_PACKED_BEGIN`?

File: tests/unit/Makefile.am [hunk 14]
```diff
@@ -100,6 +101,9 @@ test_message_SOURCES         = test_message.cpp
 test_timer_LDADD             = $(COMMON_LDADD)
 test_timer_SOURCES           = test_timer.cpp
 
+test_toolchain_LDADD             = $(COMMON_LDADD)
```
Comment (jwhui) [inline tests/unit/Makefile.am]: Align with rest of `=`'s

**Comment by jwhui** [comment]
@turon , getting close.  See my comments inline.

File: include/platform/toolchain.h [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+ * @def OT_TOOL_PACKED_BEGIN
+ *
+ * Indicate to the compiler that a class or struct should be byte packed.
+ *
+ * Usage:
+ *
+ *    @code
+ *
+ *    typedef
+ *    OT_TOOL_PACKED_BEGIN
```
Comment (jwhui) [inline include/platform/toolchain.h]: Match OpenThread code style.

File: include/platform/toolchain.h [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+ *    struct {
+ *       char field1;
+ *       union {
+ *          char field2;
+ *          long field2;
+ *       } OT_TOOL_PACKED_FIELD;
+ *    } OT_TOOL_PACKED_END packed_struct_t;
+ *
+ *    @endcode
```
Comment (jwhui) [inline include/platform/toolchain.h]: Should this usage also be included in `OT_TOOL_PACKED_FIELD` and `OT_TOOL_PACKED_END`.

File: include/platform/toolchain.h [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @def OT_TOOL_PACKED_BEGIN
+ *
+ * Indicate to the compiler that a class or struct should be byte packed.
```
Comment (jwhui) [inline include/platform/toolchain.h]: This is not required by all compilers.  gcc, for example, won't use this macro.  Maybe better to say something like: "Compiler-specific indication that the following class or struct must be byte packed."

File: include/platform/toolchain.h [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+ *
+ * Indicate to the compiler a nested struct or union to be packed
+ * within byte packed class or struct.
+ *
+ */
+
+/**
+ * @def OT_TOOL_PACKED_END
+ *
+ * Indicate to the compiler the end of a byte packed class or struct.
```
Comment (jwhui) [inline include/platform/toolchain.h]: Maybe: "Compiler-specific indication at the end of a byte packed class or struct."

File: include/platform/toolchain.h (hunk 16 above)
Comment (turon) [inline include/platform/toolchain.h]: All the examples show up at the bottom of the same HTML page for the toolchain module, so it seems redundant unless the examples are different.

Comment (jwhui) [inline include/platform/toolchain.h]: The point is that it falls under the `OT_TOOL_PACKED_BEGIN` header.  If you want, create a `Overview` or `Summary` section separately.

Comment (turon) [inline include/platform/toolchain.h]: Good point - will do.

File: include/platform/toolchain.h [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+#elif defined(_MSC_VER)
+
+#define OT_TOOL_PACKED_BEGIN                __pragma(pack(push,1))
+#define OT_TOOL_PACKED_FIELD
+#define OT_TOOL_PACKED_END                  __pragma(pack(pop))
+#define OT_TOOL_DEPRECATED(symbol)          __pragma(deprecated(symbol))
+
+#elif defined(__SDCC)
+
+// 8051 is always byte aligned and packed
```
Comment (jwhui) [inline include/platform/toolchain.h]: s/8051/sdcc/  sdcc is not specific to 8051.

**Comment by turon** [comment]
@jwhui fixed all your latest comments

**Comment by jwhui** [comment]
@turon, any thoughts on updating the doxygen example code to the OpenThread code style?

**Comment by turon** [comment]
@jwhui Ok, I believe this toolchain PR is ready now.

**Comment by jwhui** [comment]
LGTM 👍

**Pull request merged**